File: 0155-min-stack/0155-min-stack.cpp
class MinStack {
private:
    stack<pair<int,int>> mins;
public:
    MinStack() {
    }
    
    void push(int val) {
        if(mins.empty()) {
            mins.push({val,val});
        } else {
            mins.push({val, min(mins.top().second, val)});
        }
    }
    
    void pop() {
        int temp = mins.top().second;
        mins.pop();
        if(!mins.empty()) {
            mins.top().second = max(temp, mins.top().second);
        }
    }
    
    int top() {
        return mins.top().first;
    }
    
    int getMin() {
        return mins.top().second;
    }
};

/**
 * Your MinStack object will be instantiated and called as such:
 * MinStack* obj = new MinStack();
 * obj->push(val);
 * obj->pop();
 * int param_3 = obj->top();
 * int param_4 = obj->getMin();
 */

File: 0155-min-stack/0155-min-stack.py
class MinStack:

    def __init__(self):
        self.s = deque()

    def push(self, val: int) -> None:
        if not self.s:
            self.s.append((val, val))
        else:
            top = self.s[-1]
            self.s.append((min(val, top[0]), val))

    def pop(self) -> None:
        self.s.pop()

    def top(self) -> int:
        return self.s[-1][1]

    def getMin(self) -> int:
        return self.s[-1][0]


# Your MinStack object will be instantiated and called as such:
# obj = MinStack()
# obj.push(val)
# obj.pop()
# param_3 = obj.top()
# param_4 = obj.getMin()